x-pack/plugins/maps/public/components/validated_number_input.tsx

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Component, ChangeEvent, MouseEvent } from 'react';
+import { EuiFieldNumber, EuiFormRow } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import _ from 'lodash';
+
+interface State {
+  value: number | string;
+  errorMessage: string;
+  isValid: boolean;
+}
+
+interface Props {
+  initialValue: number;
+  min: number;
+  max: number;
+  onChange: (value: number) => void;
+  label: string;
+}
+
+function getErrorMessage(min: number, max: number): string {
+  return i18n.translate('xpack.maps.validatedNumberInput.invalidClampErrorMessage', {
+    defaultMessage: 'Must be between {min} and {max}',
+    values: {
+      min,
+      max,
+    },
+  });
+}
+
+function isNumberValid(value: number | string, min: number, max: number) {
+  const parsedValue = parseFloat(value.toString());
+
+  if (isNaN(parsedValue)) {
+    return {
+      isValid: false,
+      errorMessage: getErrorMessage(min, max),
+    };
+  }
+
+  const isValid = parsedValue >= min && parsedValue <= max;
+  return {
+    parsedValue,
+    isValid,
+    errorMessage: isValid ? '' : getErrorMessage(min, max),
+  };
+}
+
+export class ValidatedNumberInput extends Component<Props, State> {
+  // @ts-expect-error state populated by getDerivedStateFromProps
+  state: State = {};
+
+  constructor(props: Props) {
+    super(props);
+
+    const { isValid, errorMessage, parsedValue } = isNumberValid(
+      props.initialValue,
+      this.props.min,
+      this.props.max
+    );
+    this.state.value = isValid ? (parsedValue as number) : props.initialValue;
+    this.state.errorMessage = errorMessage;
+    this.state.isValid = isValid;
```

Comment:
Nit: to be consistent with other constructors, state is usually set as a single object

```
this.state = {
  value: props.value,
  errorMessage,
  isValid,
}
```